File: src/ds_agent/application/dtos/work_object.py
"""DTOs for work object and integration workflow boundaries."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from ds_agent.domain.entities.external_reference import ExternalSystem
from ds_agent.domain.entities.integration_event import IntegrationEvent
from ds_agent.domain.entities.work_object import WorkObject, WorkObjectPhase


class CreateWorkObjectDTO(BaseModel):
    """Input payload for creating a work object tied to a task contract."""

    task_contract_id: str
    title: str
    request_source: str
    requestor_id: str
    requestor_display: str
    original_text: str
    channel: str | None = None
    request_metadata: dict[str, Any] = Field(default_factory=dict)
    external_reference: dict[str, Any] | None = None
    delivery_pack_id: str | None = None
    owner_agent: str = "ds-agent"
    tags: list[str] = Field(default_factory=list)
    parent_work_object_id: str | None = None


class LinkExternalReferenceDTO(BaseModel):
    """Input payload for linking an external resource to a work object."""

    work_object_id: str
    system: ExternalSystem
    resource_type: str
    resource_id: str
    url: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    idempotency_key: str | None = None
    location: Literal["request", "documentation", "follow_up"] = "documentation"
    action_type: Literal["ticket", "calendar", "message", "dashboard_update"] = "message"
    description: str | None = None
    policy_decision_id: str | None = None


class AdvanceWorkObjectPhaseDTO(BaseModel):
    """Input payload for advancing a work object's lifecycle phase."""

    work_object_id: str
    to_phase: WorkObjectPhase
    run_id: str | None = None


class CloseWorkObjectDTO(BaseModel):
    """Input payload for closing a work object."""

    work_object_id: str
    reason: str


class ListWorkObjectsDTO(BaseModel):
    """Query filters for work object listing."""

    session_id: str | None = None
    task_contract_id: str | None = None
    phase_filter: list[WorkObjectPhase] = Field(default_factory=list)
    limit: int = Field(default=20, ge=1, le=200)


class WorkObjectListItemDTO(BaseModel):
    """Compact view of a work object."""

    work_object_id: str
    task_contract_id: str
    title: str
    phase: str
    updated_at: str
    reference_count: int
    follow_up_count: int

    @classmethod
    def from_work_object(cls, work_object: WorkObject) -> WorkObjectListItemDTO:
        return cls(
            work_object_id=work_object.work_object_id,
            task_contract_id=work_object.execution.task_contract_id,
            title=work_object.title,
            phase=work_object.execution.current_phase.value,
            updated_at=work_object.updated_at.isoformat(),
            reference_count=len(work_object.documentation.references),
            follow_up_count=len(work_object.follow_up.actions),
        )


class WorkObjectViewDTO(BaseModel):
    """Detailed work object view for tools and UI surfaces."""

    work_object: WorkObject
    timeline: list[IntegrationEvent] = Field(default_factory=list)

File: electron/src/renderer/types/workObject.ts
export type WorkObjectPhase =
  | 'intake'
  | 'executing'
  | 'review'
  | 'documenting'
  | 'followup'
  | 'closed'
  | 'failed';

export type FollowUpActionStatus = 'pending' | 'completed' | 'cancelled';

export type IntegrationEventStatus = 'pending' | 'success' | 'failed' | 'dlq' | 'duplicate';

export interface ExternalReferenceView {
  system: string;
  resource_type: string;
  resource_id: string;
  url?: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
  idempotency_key: string;
}

export interface RequestSectionView {
  source: string;
  requestor_id: string;
  requestor_display: string;
  original_text: string;
  channel?: string | null;
  received_at: string;
  external_ref?: ExternalReferenceView | null;
  metadata: Record<string, unknown>;
}

export interface ExecutionSectionView {
  task_contract_id: string;
  run_ids: string[];
  current_phase: WorkObjectPhase;
  started_at?: string | null;
  completed_at?: string | null;
}

export interface DocumentationSectionView {
  delivery_pack_id?: string | null;
  references: ExternalReferenceView[];
}

export interface FollowUpActionView {
  action_type: 'ticket' | 'calendar' | 'message' | 'dashboard_update';
  description: string;
  external_ref: ExternalReferenceView;
  policy_decision_id?: string | null;
  status: FollowUpActionStatus;
}

export interface FollowUpSectionView {
  actions: FollowUpActionView[];
}

export interface WorkObjectViewModel {
  work_object_id: string;
  title: string;
  request: RequestSectionView;
  execution: ExecutionSectionView;
  documentation: DocumentationSectionView;
  follow_up: FollowUpSectionView;
  created_at: string;
  updated_at: string;
  owner_agent: string;
  tags: string[];
  parent_work_object_id?: string | null;
  metadata: Record<string, unknown>;
}

export interface WorkObjectListItemView {
  work_object_id: string;
  task_contract_id: string;
  title: string;
  phase: WorkObjectPhase;
  updated_at: string;
  reference_count: number;
  follow_up_count: number;
}

export interface IntegrationEventView {
  event_id: string;
  work_object_id: string;
  system: string;
  action: string;
  request_payload_hash: string;
  idempotency_key: string;
  status: IntegrationEventStatus;
  external_ref?: ExternalReferenceView | null;
  attempt: number;
  latency_ms: number;
  started_at: string;
  finished_at?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  policy_decision_id?: string | null;
}

export interface WorkObjectDetailView {
  work_object: WorkObjectViewModel;
  timeline: IntegrationEventView[];
}
